lib/optimizely/config_manager/http_project_config_manager.rb

```diff
@@ -0,0 +1,223 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+require_relative '../config/datafile_project_config'
+require_relative '../error_handler'
+require_relative '../exceptions'
+require_relative '../helpers/constants'
+require_relative '../logger'
+require_relative '../notification_center'
+require_relative '../project_config'
+require_relative 'project_config_manager'
+require_relative 'async_scheduler'
+require 'httparty'
+require 'json'
+module Optimizely
+  class HTTPProjectConfigManager < ProjectConfigManager
+    # Config manager that polls for the datafile and updated ProjectConfig based on an update interval.
+
+    attr_reader :config
+
+    # Initialize config manager. One of sdk_key or url has to be set to be able to use.
+    #
+    # sdk_key - Optional string uniquely identifying the datafile. It's required unless a URL is passed in.
+    # datafile: Optional JSON string representing the project.
+    # polling_interval - Optional floating point number representing time interval in seconds
+    #                  at which to request datafile and set ProjectConfig.
+    # blocking_timeout -
+    # auto_update -
+    # start_by_default -
+    # url - Optional string representing URL from where to fetch the datafile. If set it supersedes the sdk_key.
+    # url_template - Optional string template which in conjunction with sdk_key
+    #               determines URL from where to fetch the datafile.
+    # logger - Provides a logger instance.
+    # error_handler - Provides a handle_error method to handle exceptions.
+    # skip_json_validation - Optional boolean param which allows skipping JSON schema
+    #                       validation upon object invocation. By default JSON schema validation will be performed.
+    def initialize(
+      sdk_key: nil,
+      url: nil,
+      url_template: nil,
+      polling_interval: nil,
+      blocking_timeout: nil,
+      auto_update: true,
+      start_by_default: true,
+      datafile: nil,
+      logger: nil,
+      error_handler: nil,
+      skip_json_validation: false,
+      notification_center: nil
+    )
+      @logger = logger || NoOpLogger.new
+      @error_handler = error_handler || NoOpErrorHandler.new
+      @datafile_url = get_datafile_url(sdk_key, url, url_template)
+      @polling_interval = nil
+      polling_interval(polling_interval)
+      blocking_timeout ||= Helpers::Constants::CONFIG_MANAGER['BLOCKING_TIMEOUT']
+      @blocking_timeout = blocking_timeout
+      @last_modified = nil
+      @async_scheduler = AsyncScheduler.new(method(:fetch_datafile), @polling_interval, auto_update, @logger)
```

Comment:
I would suggest break `fetch_datafile` into two pieces. 
1 - Only fetch datafile and parse it and return ProjectConfig. 
2 - NotificationCenter and ProjectConfig checking, either version is same or not should be done in SetConfig. 
These two methods should come into one method and that should be passed to `AsyncScheduler`